storage: Check the txn deadline for 1PC requests

kkaneda opened the pull request:
Change `isOnePhaseCommit` to check if the transaction exceeded its deadline. If so, return false so that we do not apply the 1PC code path.

We can also just return a `TransactionAbortedError` from `executeWriteBatch`, I think.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6022)

<!-- Reviewable:end -->

kkaneda committed:
storage: Check the txn deadline for 1PC requests

Change isOnePhaseCommit to check if the transaction exceeded its
deadline. If so, return false so that we do not apply the 1PC code path.

We can also just return a TransactionAbortedError from executeWriteBatch, I think.

tschottdorf commented:
LGTM, thanks for catching this. I wonder if we can somehow factor out the preconditions for a commit into code shared by `EndTransaction` and the 1PC commit path.

tschottdorf commented on storage/replica.go:
```diff
@@ -1755,7 +1755,12 @@ func isOnePhaseCommit(ba roachpb.BatchRequest) bool {
+		return false
+	}
+	etArg := arg.(*roachpb.EndTransactionRequest)
+	if etArg.Deadline != nil && etArg.Deadline.Less(ba.Timestamp) {
```
you could just `return etArg.Deadline == nil || !etArg.Deadline.Less(ba.Timestamp))`. The other commit path also allows `ba.Timestamp == etArg.Deadline`, correct?

kkaneda replied:
Agreed. Pushed another commit to refactor the preconditions. The commit defines  `isEndTransactionTriggeringRetryError` and `isEndTransactionExceedingDeadline` and uses them in `EndTransaction` and `isOnePhaseCommit`. It's a bit weird to have two methods, but more cleanup wasn't easy.. 

Could you take a look?

kkaneda committed:
storage: Factor out EndTransaction preconditions

Define isEndTransactionTriggeringRetryError and
isEndTransactionExceedingDeadline. Use them in EndTransaction and
isOnePhaseCommit.

tschottdorf commented:
That's better, thanks. The best way would be having a single check function that returns either "ok" or a list of violated conditions... but let's not go overboard here (but we probably should the next time we touch `EndTransaction` nontrivially). Thanks and <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 1 of 2 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6022#-:-KFFQfFJqO_FPROyMhxz:-562974478)*
<!-- Sent from Reviewable.io -->

bdarnell commented on storage/replica.go:
```diff
@@ -1743,22 +1743,23 @@ func (r *Replica) executeWriteBatch(
 	}
-	return true
+	etArg := arg.(*roachpb.EndTransactionRequest)
+	return !isEndTransactionExceedingDeadline(ba.Header, *etArg)
```
This changes behavior downstream of raft and could cause replicas to diverge. On the other hand, transaction deadlines don't appear to be used outside of tests so I guess it should be fine.

kkaneda replied:
Hmm, interesting. I thought the behavior is still deterministic and won't cause replica divergence since we just compares `Deadline` in the request and the header timestamp.

bdarnell replied:
The divergence happens when the command is proposed and applied on some but not all replicas before everything is shut down for the upgrade. Then the nodes come back up with the new version and the lagging replica catches up. That replica applies the old command with the new code, producing different results.

kkaneda replied:
Aah, I see. Thanks for the explanation! Hmm, so I guess it's ok to merge for this case? Will be careful about such backward compatibility issues..

bdarnell replied:
Yeah, it's fine to merge since nothing is using this. (Spencer's change that just introduced this function also would have been unsafe to merge if deadlines were in use)

kkaneda replied:
Cool, merging then!
